### PR TITLE
Fix transient union-find, improve perf, add benchmarks and tests

### PR DESCRIPTION
**Pull request by pangloss**
I also upgraded to test against Clojure 1.6 and didn't test against older versions.

### CODE REVIEW

**Comment by jordanlewis** [comment]
Heh - whoops - we crossed workstreams. I just pushed similar fixes to master. I'll check out what you did as well.

File: .nrepl-port [hunk 1]
```diff
@@ -0,0 +1 @@
+56607
```
Comment (jordanlewis) [inline .nrepl-port]: Can you .gitignore the .nrepl-port please and remove this from the commit?

**Comment by jordanlewis** [comment]
Thanks! Let me know your thoughts about my comments. I think we should merge this once you respond to the comments and then merge to master, which means updating the semantics of count and get-canonical. Then I'll cut 0.2.

**Comment by pangloss** [comment]
Thinking about it more, I don't see a reason to have 2 types of UFNode. As long as they are only actually mutated when the UF is transient, it wouldn't matter if the mutability was always available. Any objections?

**Comment by jordanlewis** [comment]
That seems reasonable. You'll just need to make sure that you make defensive node copies correctly. If the node hasn't been copied yet, you'll have to copy it. If it's already been copied, you can mutate directly. It looks like you manage that bit of state right now by the type of the node. I think you'd need to add a field or some other way of keeping track of that state if you collapse the two types.

**Comment by pangloss** [comment]
You're right, so I'll keep the distinction. In fact I realized that if you were to do  `(let [a (-> (union-find) (conj 1 2 3) transient (union! 1 2) persistent!) b (-> transient (union 1 3))] [a b]` the `a` version  could be impacted by changes in the `b` versien. I fixed that in the previous commit.

I think this is ready to go.

**Comment by jordanlewis** [comment]
Cool. That makes sense. We should probably add tests for that case that you mentioned.

One more annoying question - why should we bother making UFNode mutable in the first place? Surely the performance gains achieved by mutating a UFNode vs just allocating a new one are minimal if existent, right? It's not like it's a whole tree - it's just a record with a few shallow fields. Do you think it's worth the added complexity of managing the two types of node with their mutable flags?

**Comment by pangloss** [comment]
By my benchmark it tripled the speed. I'll add the test.

**Comment by jordanlewis** [comment]
Sweet!

**Comment by pangloss** [comment]
Ok, that's done.

**Comment by jordanlewis** [comment]
Thanks!

**Pull request merged**